lib/src/solver/incompatibility.dart

```diff
@@ -44,39 +52,84 @@ class Incompatibility {
       }
     }
 
-    return new Incompatibility._(byName.values.expand((byRef) {
-      // If there are any positive terms for a given package, we can discard
-      // any negative terms.
-      var positiveTerms =
-          byRef.values.where((term) => term.isPositive).toList();
-      if (positiveTerms.isNotEmpty) return positiveTerms;
+    return new Incompatibility._(
+        byName.values.expand((byRef) {
+          // If there are any positive terms for a given package, we can discard
+          // any negative terms.
+          var positiveTerms =
+              byRef.values.where((term) => term.isPositive).toList();
+          if (positiveTerms.isNotEmpty) return positiveTerms;
 
-      return byRef.values;
-    }).toList());
+          return byRef.values;
+        }).toList(),
+        cause);
   }
 
-  Incompatibility._(this.terms);
+  Incompatibility._(this.terms, this.cause);
 
   String toString() {
+    if (cause == IncompatibilityCause.dependency) {
```

Comment:
What do you think of moving most of this logic into IncompatibilityCause? Something like:

```dart
class Incompability {
  String toString() => cause.describe(terms);
}

abstract class IncompatibilityCause {
  String describe(List<Term> terms);
}

class SdkCause implements IncompatibilityCause {
  String describe(List<Term> terms) {
    // ...
  }
}
```

?

Reply:
It always weirds me out a bit to have a method like `Incompatibility.describe()` that's only ever supposed to be invoked by one specific caller... it feels like kind of the wrong separation of concerns.